spec/models/principals/scopes/visible_spec.rb

```diff
@@ -59,7 +59,7 @@
       current_user { create :user, global_permissions: %i[manage_user] }
 
       it 'sees all users' do
-        expect(subject).to match_array [current_user, other_project_user, global_user]
+        expect(subject).to match_array [current_user, other_project_user, global_user, User.anonymous]
```

Comment:
> The expectation is correct given the implementation. But I don't understand where the anonymous user originates from in the setup. If that is done somehow by e.g. project creation, I'd still welcome the anonymous user creation to be listed explicitly as e.g. a shared_let.

Yes, I can add a shared_let, or explicitly set the current user when the project gets created.

The thing that bothers me, and maybe you can help, is that in the second test, the user from the other project does not see the anonymous user. I do not understand why.

> Since Principal.all are returned effectively, the other principals like a group, a placeholder user and the SystemUser should also be covered. Can be part of a different PR, though.

You mean we should add more tests to ensure groups, placeholders, and systemuser are correctly returned when `Principal.visible` is called?
If so I can tackle that in another PR as you suggested.

Reply:
> The thing that bothers me, and maybe you can help, is that in the second test, the user from the other project does not see the anonymous user. I do not understand why.

The reason for this is that with the permissions set of the set up for the second test, the tested for users has neither 
`manage_members` nor `manage_user`. In that case, only users, with a membership in a project the user is also a member in are returned. This explains, why the AnonymousUser is not returned, since that never actually becomes a member, even in public projects. That user is the fallback, that not logged in users are mapped to if non authenticated access is allowed. 

What I find more surprising about the second test case is that the user is actually allowed to see the `other_project_user` since the queried for user does not have the `view_members` permission in the project `other_project`. My explanation for this would be that the two users need to see each other in order to interact with each other. And the user does not know which project the other user is a member of (if there were more than one project).

Reply:
48d33595c591e4a1294048d3140c548decb1137f

Reply:
> Since Principal.all are returned effectively, the other principals like a group, a placeholder user and the SystemUser should also be covered. Can be part of a different PR, though.

Added in PR #11757